File: tests/build-index.ts
#!/usr/bin/env node
/**
 * Regenerate skills/INDEX.md from current skills/<name>/SKILL.md frontmatter.
 *
 * Usage: tsx tests/build-index.ts  (or  npm run index)
 */
import { readdir, readFile, writeFile, stat } from "node:fs/promises";
import { resolve, dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = resolve(__dirname, "..");
const SKILLS_DIR = resolve(ROOT, "skills");
const INDEX = join(SKILLS_DIR, "INDEX.md");

async function getDescription(skillMd: string): Promise<string> {
  const content = await readFile(skillMd, "utf-8");
  const head = content.slice(0, 2000);
  if (head.startsWith("---")) {
    const end = head.indexOf("---", 3);
    if (end > 0) {
      for (const line of head.slice(3, end).split("\n")) {
        if (line.startsWith("description:")) {
          return line.split(":", 2)[1].trim().replace(/^["']|["']$/g, "");
        }
      }
    }
  }
  for (const line of head.split("\n")) {
    const t = line.trim();
    if (t && !t.startsWith("#") && !t.startsWith("---")) return t.slice(0, 80);
  }
  return "";
}

async function collectRows(dir: string, prefix: string): Promise<Array<[string, string]>> {
  let entries: string[] = [];
  try {
    entries = await readdir(dir);
  } catch {
    return [];
  }
  const rows: Array<[string, string]> = [];
  for (const s of entries.sort()) {
    // skip the archive subdir (collected separately) and non-dir entries like INDEX.md
    if (s === "archive") continue;
    if (!(await stat(join(dir, s)).then((st) => st.isDirectory()).catch(() => false))) continue;
    const sm = join(dir, s, "SKILL.md");
    const desc = (await getDescription(sm).catch(() => "(missing SKILL.md)")) || "(missing SKILL.md)";
    rows.push([prefix + s, desc]);
  }
  return rows;
}

async function main(): Promise<number> {
  const active = await collectRows(SKILLS_DIR, "");
  const archived = await collectRows(join(SKILLS_DIR, "archive"), "archive/");
  const rows = [...active, ...archived];
  let out = `# Skill Index\n\nAuto-generated. Re-run with: npm run index\nTotal: ${active.length} active + ${archived.length} archived = ${rows.length} skills\n\n`;
  out += `## Active (${active.length})\n\n| Skill | Description |\n|-------|-------------|\n`;
  for (const [s, d] of active) {
    const dSafe = d.replace(/\|/g, "\\|").slice(0, 100);
    out += `| ${s} | ${dSafe} |\n`;
  }
  out += `\n## Archived (${archived.length})\n\n> Not in active use. Restore via: \`git mv skills/archive/<name> skills/<name>\`\n\n| Skill | Description |\n|-------|-------------|\n`;
  for (const [s, d] of archived) {
    const dSafe = d.replace(/\|/g, "\\|").slice(0, 100);
    out += `| ${s} | ${dSafe} |\n`;
  }
  await writeFile(INDEX, out, "utf-8");
  console.log(`wrote ${INDEX} (${active.length} active, ${archived.length} archived)`);
  return 0;
}

process.exit(await main());

File: tests/build-index.py
#!/usr/bin/env python3
"""Regenerate skills/INDEX.md from current skills/<name>/SKILL.md frontmatter.

Usage: python3 tests/build-index.py
"""
import os
import sys

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SKILLS_DIR = os.path.join(ROOT, "skills")
INDEX = os.path.join(SKILLS_DIR, "INDEX.md")


def get_description(skill_md: str) -> str:
    content = open(skill_md, "r", encoding="utf-8").read(2000)
    if content.startswith("---"):
        end = content.find("---", 3)
        if end > 0:
            for line in content[3:end].split("\n"):
                if line.startswith("description:"):
                    return line.split(":", 1)[1].strip().strip('"').strip("'")
    for line in content.split("\n"):
        line = line.strip()
        if line and not line.startswith("#") and not line.startswith("---"):
            return line[:80]
    return ""


def main() -> int:
    if not os.path.isdir(SKILLS_DIR):
        print(f"no skills/ dir at {SKILLS_DIR}")
        return 1
    skill_dirs = sorted(
        e for e in os.listdir(SKILLS_DIR)
        if os.path.isdir(os.path.join(SKILLS_DIR, e))
    )
    rows = []
    for s in skill_dirs:
        sm = os.path.join(SKILLS_DIR, s, "SKILL.md")
        desc = get_description(sm) if os.path.exists(sm) else "(missing SKILL.md)"
        rows.append((s, desc))
    with open(INDEX, "w", encoding="utf-8") as f:
        f.write("# Skill Index\n\n")
        f.write(f"Auto-generated. Total: {len(rows)} skills.\n\n")
        f.write("| Skill | Description |\n")
        f.write("|-------|-------------|\n")
        for s, d in rows:
            d_safe = d.replace("|", "\\|")[:100]
            f.write(f"| {s} | {d_safe} |\n")
    print(f"wrote {INDEX} ({len(rows)} skills)")
    return 0


if __name__ == "__main__":
    sys.exit(main())
